src/Controller/CRUDController.php

```diff
@@ -66,11 +68,29 @@ class CRUDController extends Controller
      */
     private $templateRegistry;
 
-    public function setContainer(?ContainerInterface $container = null): void
+    public function setContainer(ContainerInterface $container): ?ContainerInterface
     {
-        $this->container = $container;
+        $result = parent::setContainer($container);
 
         $this->configure();
+
+        return $result;
+    }
+
+    public static function getSubscribedServices()
+    {
+        return array_merge(parent::getSubscribedServices(), [
+            'sonata.admin.pool' => Pool::class,
+            'sonata.admin.breadcrumbs_builder' => BreadcrumbsBuilderInterface::class,
+            'sonata.admin.audit.manager' => AuditManagerInterface::class,
+            'sonata.admin.object.manipulator.acl.admin' => AdminObjectAclManipulator::class,
+            'sonata.exporter.exporter' => '?'.Exporter::class,
+            'sonata.admin.admin_exporter' => '?'.AdminExporter::class,
+            'sonata.admin.security.acl_user_manager' => '?object',
```

Comment:
not sure this works using object...

Reply:
No idea, but based on how this is used:

https://github.com/sonata-project/SonataAdminBundle/blob/51ca7b76aa9104cd14b435906eb4156de96d3ca7/src/Controller/CRUDController.php#L1358-L1365

Maybe we could create an interface with the `findUsers` method, provide integration with fos user manager and trigger a deprecation in case the acl manager passed does not implement the interface.

https://github.com/sonata-project/SonataAdminBundle/blob/e51dafec316b94aeda41d4646837c1a26c869393/src/DependencyInjection/SonataAdminExtension.php#L118-L122

Reply:
I agree that an interface could be great to help the developper knowing which methods should have a `userManager`.

Reply:
This is already done on 3.x